integration_tests/src/lib.rs

```diff
@@ -168,21 +502,25 @@ mod test {
             // If start_downstairs returned Ok, then address will be populated
             self.downstairs.as_ref().unwrap().repair_address()
         }
+
+        pub fn snapshot_exists(&self, snapshot_name: String) -> Result<bool> {
+            self.dataset.snapshot_exists(snapshot_name)
+        }
     }
 
     #[derive(Debug)]
-    struct TestDownstairsSet {
-        downstairs1: TestDownstairs,
-        downstairs2: TestDownstairs,
-        downstairs3: TestDownstairs,
+    struct TestDownstairsSet<T: TestDownstairsDataset> {
+        downstairs1: TestDownstairs<T>,
+        downstairs2: TestDownstairs<T>,
+        downstairs3: TestDownstairs<T>,
         crucible_opts: CrucibleOpts,
         blocks_per_extent: u64,
         extent_count: u32,
     }
 
-    impl TestDownstairsSet {
+    impl<T: TestDownstairsDataset> TestDownstairsSet<T> {
         /// Spin off three downstairs, with a 5120b region
-        pub async fn small(read_only: bool) -> Result<TestDownstairsSet> {
+        pub async fn small(read_only: bool) -> Result<TestDownstairsSet<T>> {
```

Comment:
Optional: you could make these functions return `Result<Self>` to avoid repeating `<T>`